### ISSUE TITLE
ADTs with hole type consistency have casting bug

### ISSUE THREAD

**Issue by gensofubi** [hazelgrove/hazel#1311]
<img width="371" alt="601521e61192a0b8ae8a22ad3219bcab" src="https://github.com/hazelgrove/hazel/assets/71319371/35e9a500-60a1-4565-95bf-b3fbe46ff077">
<img width="371" alt="67113d16c739bb9f9c84cbfb6a887b06" src="https://github.com/hazelgrove/hazel/assets/71319371/612457ef-ab45-4be5-9aa8-c99b087b57f6">

I expect both examples to work alright but one gives exceptions and the other is stucked.

**Comment by Negabinary** [comment]
I just now pushed a quick fix to get this example working on that branch (https://[hazel](https://hazel.org/build/remove-dhexp/).org/build/remove-dhexp/). 

I've changed a lot of how elaboration and pattern matching work on the #1197 branch. I think if we want to fix it on dev, it'll be somewhere in the matches cast sum function in PatternMatch.re. Hopefully #1197 will get merged soon though, so we may not need to do it on dev?
